--- Day 7: Camel Cards ---
Your all-expenses-paid trip turns out to be a one-way, five-minute ride in an airship. (At least it's a cool airship!) It drops you off at the edge of a vast desert and descends back to Island Island.

"Did you bring the parts?"

You turn around to see an Elf completely covered in white clothing, wearing goggles, and riding a large camel.

"Did you bring the parts?" she asks again, louder this time. You aren't sure what parts she's looking for; you're here to figure out why the sand stopped.

"The parts! For the sand, yes! Come with me; I will show you." She beckons you onto the camel.

After riding a bit across the sands of Desert Island, you can see what look like very large rocks covering half of the horizon. The Elf explains that the rocks are all along the part of Desert Island that is directly above Island Island, making it hard to even get there. Normally, they use big machines to move the rocks and filter the sand, but the machines have broken down because Desert Island recently stopped receiving the parts they need to fix the machines.

You've already assumed it'll be your job to figure out why the parts stopped when she asks if you can help. You agree automatically.

Because the journey will take a few days, she offers to teach you the game of Camel Cards. Camel Cards is sort of similar to poker except it's designed to be easier to play while riding a camel.

In Camel Cards, you get a list of hands, and your goal is to order them based on the strength of each hand. A hand consists of five cards labeled one of A, K, Q, J, T, 9, 8, 7, 6, 5, 4, 3, or 2. The relative strength of each card follows this order, where A is the highest and 2 is the lowest.

Every hand is exactly one type. From strongest to weakest, they are:

Five of a kind, where all five cards have the same label: AAAAA
Four of a kind, where four cards have the same label and one card has a different label: AA8AA
Full house, where three cards have the same label, and the remaining two cards share a different label: 23332
Three of a kind, where three cards have the same label, and the remaining two cards are each different from any other card in the hand: TTT98
Two pair, where two cards share one label, two other cards share a second label, and the remaining card has a third label: 23432
One pair, where two cards share one label, and the other three cards have a different label from the pair and each other: A23A4
High card, where all cards' labels are distinct: 23456
Hands are primarily ordered based on type; for example, every full house is stronger than any three of a kind.

If two hands have the same type, a second ordering rule takes effect. Start by comparing the first card in each hand. If these cards are different, the hand with the stronger first card is considered stronger. If the first card in each hand have the same label, however, then move on to considering the second card in each hand. If they differ, the hand with the higher second card wins; otherwise, continue with the third card in each hand, then the fourth, then the fifth.

So, 33332 and 2AAAA are both four of a kind hands, but 33332 is stronger because its first card is stronger. Similarly, 77888 and 77788 are both a full house, but 77888 is stronger because its third card is stronger (and both hands have the same first and second card).

To play Camel Cards, you are given a list of hands and their corresponding bid (your puzzle input). For example:

32T3K 765
T55J5 684
KK677 28
KTJJT 220
QQQJA 483
This example shows five hands; each hand is followed by its bid amount. Each hand wins an amount equal to its bid multiplied by its rank, where the weakest hand gets rank 1, the second-weakest hand gets rank 2, and so on up to the strongest hand. Because there are five hands in this example, the strongest hand will have rank 5 and its bid will be multiplied by 5.

So, the first step is to put the hands in order of strength:

32T3K is the only one pair and the other hands are all a stronger type, so it gets rank 1.
KK677 and KTJJT are both two pair. Their first cards both have the same label, but the second card of KK677 is stronger (K vs T), so KTJJT gets rank 2 and KK677 gets rank 3.
T55J5 and QQQJA are both three of a kind. QQQJA has a stronger first card, so it gets rank 5 and T55J5 gets rank 4.
Now, you can determine the total winnings of this set of hands by adding up the result of multiplying each hand's bid with its rank (765 * 1 + 220 * 2 + 28 * 3 + 684 * 4 + 483 * 5). So the total winnings in this example are 6440.

Find the rank of every hand in your set. What are the total winnings?

In [ ]:
from collections import Counter

def hand_type(hand):
    """
    returns type scoring:
    50 for five of a kind
    40 for 4 of a kind
    31 for full house
    30 for trips
    21 for two pair
    20 for single pair
    10 else (high card)
    """
    hand = Counter(hand)
    score = hand.most_common(1)[0][1] * 10
    if 20 <= score <= 30:
        if hand.most_common()[1][1] == 2:
            score += 1
    return score

def card_score(hand):
    """
    scores 2-10 for numbered cards, 11 J, 12 Q, 13 K, 14 A
    """
    scores = []
    for card in hand:
        if card == 'T':
            score = 10
        elif card == 'J':
            score = 11
        elif card == 'Q':
            score = 12
        elif card == 'K':
            score = 13
        elif card == 'A':
            score = 14
        else:
            score = int(card)
        scores.append(score)
    return tuple(scores)

hand_type('QQQJJ'), card_score('QQQ9T')

(31, (12, 12, 12, 9, 10))

In [19]:
with open('input07.txt') as file:
    data = [(hand_type(line[:5]), card_score(line[:5]), line[:5], int(line[6:])) for line in file] # hand type, kicker, hand, bid
data[:5]

[(20, (11, 2, 5, 6, 6), 'J2566', 131),
 (31, (13, 7, 13, 13, 7), 'K7KK7', 272),
 (40, (2, 13, 2, 2, 2), '2K222', 74),
 (21, (4, 4, 10, 5, 5), '44T55', 467),
 (40, (3, 3, 4, 3, 3), '33433', 226)]

In [20]:
data.sort()
data

[(10, (2, 3, 5, 10, 11), '235TJ', 179),
 (10, (2, 5, 8, 6, 3), '25863', 725),
 (10, (2, 5, 10, 4, 6), '25T46', 95),
 (10, (2, 6, 5, 11, 13), '265JK', 344),
 (10, (2, 7, 8, 5, 9), '27859', 72),
 (10, (2, 7, 8, 14, 5), '278A5', 395),
 (10, (2, 10, 3, 11, 12), '2T3JQ', 280),
 (10, (2, 10, 8, 11, 4), '2T8J4', 638),
 (10, (2, 10, 9, 5, 3), '2T953', 933),
 (10, (2, 11, 14, 4, 3), '2JA43', 282),
 (10, (2, 12, 8, 6, 13), '2Q86K', 693),
 (10, (2, 12, 9, 6, 10), '2Q96T', 398),
 (10, (2, 13, 8, 4, 11), '2K84J', 329),
 (10, (2, 13, 11, 7, 5), '2KJ75', 436),
 (10, (2, 14, 5, 7, 3), '2A573', 741),
 (10, (3, 2, 4, 8, 12), '3248Q', 235),
 (10, (3, 2, 4, 11, 9), '324J9', 105),
 (10, (3, 2, 14, 5, 4), '32A54', 746),
 (10, (3, 4, 6, 14, 10), '346AT', 426),
 (10, (3, 4, 12, 9, 7), '34Q97', 229),
 (10, (3, 4, 14, 11, 13), '34AJK', 769),
 (10, (3, 6, 13, 10, 9), '36KT9', 809),
 (10, (3, 7, 14, 11, 8), '37AJ8', 811),
 (10, (3, 8, 9, 2, 11), '3892J', 169),
 (10, (3, 9, 2, 12, 13), '392QK', 679),
 (10, (3, 9, 

In [21]:
total_winnings = 0
for ix, hand in enumerate(data):
    rank = ix + 1
    winnings = rank * hand[3]
    total_winnings += winnings
total_winnings

250602641

Your puzzle answer was 250602641.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
To make things a little more interesting, the Elf introduces one additional rule. Now, J cards are jokers - wildcards that can act like whatever card would make the hand the strongest type possible.

To balance this, J cards are now the weakest individual cards, weaker even than 2. The other cards stay in the same order: A, K, Q, T, 9, 8, 7, 6, 5, 4, 3, 2, J.

J cards can pretend to be whatever card is best for the purpose of determining hand type; for example, QJJQ2 is now considered four of a kind. However, for the purpose of breaking ties between two hands of the same type, J is always treated as J, not the card it's pretending to be: JKKK2 is weaker than QQQQ2 because J is weaker than Q.

Now, the above example goes very differently:

32T3K 765
T55J5 684
KK677 28
KTJJT 220
QQQJA 483
32T3K is still the only one pair; it doesn't contain any jokers, so its strength doesn't increase.
KK677 is now the only two pair, making it the second-weakest hand.
T55J5, KTJJT, and QQQJA are now all four of a kind! T55J5 gets rank 3, QQQJA gets rank 4, and KTJJT gets rank 5.
With the new joker rule, the total winnings in this example are 5905.

Using the new joker rule, find the rank of every hand in your set. What are the new total winnings?



In [32]:
def hand_type(hand, joker=False):
    """
    returns type scoring:
    50 for five of a kind
    40 for 4 of a kind
    31 for full house
    30 for trips
    21 for two pair
    20 for single pair
    10 else (high card)
    """
    hand = Counter(hand)
    if joker:
        j = hand['J']
        hand['J'] = 0
    score = hand.most_common(1)[0][1] * 10
    if 20 <= score <= 30:
        if hand.most_common()[1][1] == 2:
            score += 1
    if joker:
        score += j * 10
    return score
hand_type('JTTQQ', joker=True)

31

In [34]:
def card_score(hand, joker=False):
    """
    scores 2-10 for numbered cards, 11 J, 12 Q, 13 K, 14 A
    if joker: J scores 1
    """
    scores = []
    for card in hand:
        if card == 'T':
            score = 10
        elif card == 'J':
            if joker:
                score = 1
            else:
                score = 11
        elif card == 'Q':
            score = 12
        elif card == 'K':
            score = 13
        elif card == 'A':
            score = 14
        else:
            score = int(card)
        scores.append(score)
    return tuple(scores)

hand_type('QQQ9J',joker=True), card_score('QQQ9J',joker=True)

(40, (12, 12, 12, 9, 1))

In [38]:
data_joker = [(hand_type(hand, joker=True), card_score(hand,joker=True), hand, bid) for _, _, hand, bid in data]
data_joker.sort()

total_winnings = 0
for ix, hand in enumerate(data_joker):
    rank = ix + 1
    winnings = rank * hand[3]
    total_winnings += winnings
total_winnings

251037509

That's the right answer! You are one gold star closer to restoring snow operations.

You have completed Day 7! You can [Share] this victory or [Return to Your Advent Calendar].